### DISCUSSION TITLE
model_validate to convert between pydantic models?

### DISCUSSION THREAD

**Discussion by OliverFarren** [pydantic/pydantic#9490] [Question]
It's a common scenario where I'll want to convert from an internal representation to a presentation representation which in practice means just filtering out some internal information. As I understand it, there are few ways I could do this:

``` python
from pydantic import BaseModel, ConfigDict

class ProjectInternal(BaseModel):
    name: str
    project_id: str


class ProjectA(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str


class ProjectB(BaseModel):
    name: str

project_internal = ProjectInternal(name="Hello", project_id="foo")

# Method A - from attributes
project_a = ProjectA.model_validate(project_internal)

# Method B - serialise and deserialise
project_b1 = ProjectB.model_validate(project_internal.model_dump())
project_b2 = ProjectB(**project_internal.model_dump())  # slightly different style

```

As far as I can see both approaches are fine, from testing this toy example with `timeit` several times the time difference was between 17% - 35% faster using Method A, given the variance I suspect that's mostly noise and there isn't going to be much in it performance wise. But then again, maybe with more complex models one is better.

So as a point of discussion i'm wondering if there's a reason I might consider one approach over the other? specifically in the context of converting between pydantic models. I assume there's good reason that `from_attributes` isn't the defacto method.

---

N.B. for this example and discussion i'm using pydantic==2.7.1, pydantic-core==2.18.2, Python 3.11.1

**Comment by OliverFarren** [comment]
Perhaps in answer to my own discussion, the `from_attributes` method doesn't work for nested models which makes sense, so I suppose Method B is the more flexible approach to take

``` python
from pydantic import BaseModel, ConfigDict


class EntityInternal(BaseModel):
    id: int
    name: str


class Entity(BaseModel):
    name: str


class ProjectInternal(BaseModel):
    name: str
    project_id: str
    entities: list[EntityInternal]


class ProjectA(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str
    entities: list[Entity]


class ProjectB(BaseModel):
    name: str
    entities: list[Entity]



project_internal = ProjectInternal.model_validate(
    {
        "name": "Hello",
        "project_id": "foo",
        "entities": [{"id": idx, "name": name} for idx, name in enumerate('abcdefg')]
    }
)


# Method A - from attributes
#project_a = ProjectA.model_validate(project_internal)  # Validation Error

# Method B - serialise and deserialise
project_b1 = ProjectB.model_validate(project_internal.model_dump())
project_b2 = ProjectB(**project_internal.model_dump())  # slightly different style
```

**Reply by d812k** [reply]
Using the `ConfigDict(from_attributes=True)` works perfectly fine for the nested models as well and this way is mentioned in the pydantic documentation https://docs.pydantic.dev/latest/concepts/models/#nested-attributes What's missing from the code above is `model_config = ConfigDict(from_attributes=True)` in the internal `Entity` model.
